### DISCUSSION TITLE
types in completion menu

### DISCUSSION THREAD

**Discussion by lewis6991** [scalameta/nvim-metals#241] [Q&A]
Just updated this after not using for a while and I've found the completion menu has become very overwhelming due to the expansion of types.

![image](https://user-images.githubusercontent.com/7904185/136187337-3a90b1ec-3717-4c70-a99c-10f1deabbaa8.png)

I'm using https://github.com/hrsh7th/nvim-cmp with the config;

```lua
local cmp = require 'cmp'
cmp.setup {
  mapping = {...},
  sources = {
    { name = 'nvim_lsp' },
    ...
  },
}

```

Any idea if it's possible to make the completion menu easier to read?

**Comment by ckipp01** [comment]
Hey @lewis6991! This is a timely discussion. This actually just came up in https://github.com/scalameta/metals-feature-requests/issues/235. Feel free to add anything there. I think this is something I'll need to take a closer look at on the Metals side to see how we can improve this.

**Reply by lewis6991** [reply]
Ah cool, I'll take a look.

Just as a quick workaround, I came up with this for nvim-cmp:

```lua
local maxwidth = 40
cmp.setup {
  formatting = {
    format = function(entry, vim_item)
      if #vim_item.abbr > maxwidth then
        vim_item.abbr = vim_item.abbr:sub(1, maxwidth)..'...'
      end
      return vim_item
    end
  }
}
```

This at least allows you to look at the documentation pop-up if you need to see the full signature.

**Reply by suliatis** [reply]
Hi @lewis6991 

you made me thinking and I came up another workaround.

```lua
  cmp.setup {
   -- ...
    formatting = {
      format = function(entry, vim_item)
        if vim_item.kind == 'Method' then
          vim_item.abbr = vim_item.word
        end
        return vim_item
      end,
    },
  }
```

Here I replace `abbr` with `word` if the `kind` is method. The `word` field represents the text that will be inserted after you confirmed the selection, so it does not contain the type signature. The only drawback is overloaded methods can't be differentiate until you hover them. And it looks like this:

<img width="2301" alt="Screen Shot 2021-10-06 at 21 21 41" src="https://user-images.githubusercontent.com/6475233/136269297-1d7d8b5d-4e15-4c7a-b6c3-f605cba40bfc.png">

To see more: `help complete-items`
